docs/architecture.md

```diff
@@ -0,0 +1,96 @@
+# Pulp CLI Architecture
+
+## Plugin system
+
+The Pulp CLI is designed with a plugin structure. Plugins can either live in the main package or be shipped independently.
+By convention, all parts of the CLI are packages in the open namespace `pulpcore.cli`.
+A plugin can register itself with the main app by specifying by specifying its main module as a `pulp_cli.plugins` entrypoint in `setup.py`.
+
+```python
+entry_points={
+    "pulp_cli.plugins": [
+        "myplugin=pulpcore.cli.myplugin",
+    ],
+}
+```
+
+The plugin can then attach subcommands to the `pulpcore.cli.common.main` command.
+
+```python
+from pulpcore.cli.common import main
+
+@main.command()
+def my_command():
+    pass
+```
+
+## Contexts
+
+In `click`, every subcommand is accompanied by a `click.Context`, and objects can be attached to them.
```

Comment:
Is this always true? If not, I'd suggest:

```suggestion
In `click`, every subcommand is usually accompanied by a `click.Context`, and objects can be attached to them.
```

Reply:
It is. The command you are writing and decorating is only called via `ctx.invoke()`.
Optional is the use of `ctx.obj`.